src/commands/login/actions.ts

```diff
@@ -14,15 +13,19 @@ export const loginWithToken = async (token: string, region: RegionCode) => {
     })
   }
   catch (error) {
-    if ((error as FetchError).response?.status === 401) {
-      throw new Error(`The token provided ${chalk.bold(maskToken(token))} is invalid: ${chalk.bold(`401 ${(error as FetchError).data.error}`)}
+    if (error instanceof FetchError) {
```

Comment:
@edodusi because it has a custom message for 401

Reply:
can't you do it directly in `handleAPIError` by using a conditional with `login_with_token`? This way it should be easier to maintain/extend

Reply:
I prefer not to make `handleApiError` flexible, the scope of that function is to handle the error with general error messages.

What if I want to pass a custom message for 422 instead?

Reply:
I think the responsibility for printing messages should be of  `handleAPIError`, and the message to pick can depend on different factors: the http status, the action, the error, even the region potentially. 

I would apply the single-responsibility principle, these actions should only handle the actual action to perform, not care about how errors are handled, that is the responsibility of another function or class.

Reply:
The responsibility of printing is not `handleAPIError`, is konsola, and konsola is being called only in the top-level `handleError` method. 

The only purpose of `handleAPIError` is not to repeat the same in each action. Completely different purposes

Reply:
We are bubbling the errors up instead of printing them on lower levels

Reply:
Ok the printing responsibilly is of `konsola` but `handleAPIError` has to choose the correct `APIError` to instantiate, it's like a factory, so in my opinion the entire responsibility of calling the correct `APIError` should be in that function.

Imagine you want to add an icon to each and every error messages (absurd idea), in this scenario you should be able to do this just by changing a single file, instead in this way you are forced to change at least two

Reply:
Sorry @edodusi I don't think I follow. 

The responsibility of `handleAPIError` is to throw an instance of `APIError` with the proper correspondent error message depending on the response status, is not a factory returning any constructor.

```ts
export function handleAPIError(action: keyof typeof API_ACTIONS, error: Error): void {
  if (error instanceof FetchError) {
    const status = error.response?.status

    switch (status) {
      case 401:
        throw new APIError('unauthorized', action, error)
      case 422:
        throw new APIError('invalid_credentials', action, error)
      default:
        throw new APIError('network_error', action, error)
    }
  }
  else if (error.message.includes('NetworkError') || error.message.includes('Failed to fetch')) {
    throw new APIError('network_error', action, error)
  }
  throw new APIError('generic', action, error)
}
```

Then this is caught by `handleError` at top level, whose responsibility is to check what type of error it is to be able to call konsola with the right prefix in case the flag verbose is included.

```ts
export function handleError(error: Error, verbose = false): void {
  // If verbose flag is true and the error has getInfo method
  if (verbose && typeof (error as any).getInfo === 'function') {
    const errorDetails = (error as any).getInfo()
    if (error instanceof CommandError) {
      konsola.error(`Command Error: ${error.getInfo().message}`, errorDetails)
    }
    else if (error instanceof APIError) {
      konsola.error(`API Error: ${error.getInfo().cause}`, errorDetails)
    }
    else if (error instanceof FileSystemError) {
      konsola.error(`File System Error: ${error.getInfo().cause}`, errorDetails)
    }
    else {
      konsola.error(`Unexpected Error: ${error.message}`, errorDetails)
    }
  }
  else {
    // Print the message stack if it exists
    if ((error as any).messageStack) {
      const messageStack = (error as any).messageStack
      messageStack.forEach((message: string, index: number) => {
        konsola.error(message, null, {
          header: index === 0,
          margin: false,
        })
      })
      konsola.br()
      konsola.info('For more information about the error, run the command with the `--verbose` flag')
    }
    else {
      konsola.error(error.message, null, {
        header: true,
      })
    }
  }

  if (!process.env.VITEST) {
    console.log('') // Add a line break for readability
    process.exit(1) // Exit process if not in a test environment
  }
}
```

> Imagine you want to add an icon to each and every error messages (absurd idea), in this scenario you should be able to do this just by changing a single file, instead in this way you are forced to change at least two

Would be enough to change the error message format once on the `konsola` file? 

```
error: (message: string, info?: unknown, options?: KonsolaFormatOptions) => {
    if (options?.header) {
      const errorHeader = chalk.bgRed.bold.white(` Error `)
      console.error(formatHeader(errorHeader))
      console.log('') // Add a line break
    }

    console.error(`${chalk.red.bold('▲ error')} ${message}`, info || '')
    if (options?.margin) {
      console.error('') // Add a line break
    }
  },
  ```

Reply:
@edodusi think of it like this:

- `handleAPIError` throw API Error depending on the HTTP response code
- `handleFilesystemError` trow FileSystemError depending on the filesysten operation code
- `handleError` top-level that catch all errors thrown in the CLI and send them to the konsola
- `konsola` prompt formatting and UI

Reply:
@alvarosabu if this last comment was true I would expect `APIError` to be raised **exclusively** by `handleAPIError`, and instead you throw it also in the action handler, that's what I'm saying.

I used the factory as an analogy as it's the class responsible for instantiating object, in this case it's responsible for throwing errors.

But the point is, and this is only because the code is well architected that this pops out to me, there is an exception to this pattern, that is that an action is throwing its custom errors instead of delegating this responsibility to `handleAPIError`.

If the code was just made of catch blocks that throw errors with custom messages and log something and do their things I would not have argued, but since you have this design pattern and I think it's good and extensible and easy to maintain, then (to me) the logical consequence is to handle everything about errors in `handleAPIError`.

You say this is an exception because it has to print its own custom message, ok, then if tomorrow we add another action we have to think where to put the custom message, instead a good design means we don't have to think, we simply follow, extend, etc...

Reply:
I see your point. To me this is NOT a blocker, probably I just have a slightly different view of the architecture of this code but we can discuss, and probably with future improvements we will better align.

Thanks for the great discussion, I love to see deep thoughts on software design ❤️ 

You can resolve this conversation, and soon as we need to extend this behavior we will think of the best way.